File: ee/tabby-db/src/attachment.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct Attachment {
    pub code: Option<Vec<AttachmentCode>>,
    pub client_code: Option<Vec<AttachmentClientCode>>,
    pub doc: Option<Vec<AttachmentDoc>>,
    pub code_file_list: Option<AttachmentCodeFileList>,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentCodeFileList {
    pub file_list: Vec<String>,

    #[serde(default)]
    pub truncated: bool,
}

/// AttachmentDoc represents a union of various document types.
/// For backward compatibility, it is stored untagged in the database.
/// Ensure that new document types have unique field names to avoid deserialization issues.
#[derive(Serialize, Deserialize)]
#[serde(untagged)] // Mark the serde serialization format as untagged for backward compatibility: https://serde.rs/enum-representations.html#untagged
pub enum AttachmentDoc {
    Web(AttachmentWebDoc),
    Issue(AttachmentIssueDoc),
    Pull(AttachmentPullDoc),
    Commit(AttachmentCommitDoc),
    Page(AttachmentPageDoc),
    Ingested(AttachmentIngestedDoc),
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentWebDoc {
    pub title: String,
    pub link: String,
    pub content: String,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentIssueDoc {
    pub title: String,
    pub link: String,
    pub author_user_id: Option<String>,
    pub body: String,
    pub closed: bool,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentPullDoc {
    pub title: String,
    pub link: String,
    pub author_user_id: Option<String>,
    pub body: String,
    pub diff: String,
    pub merged: bool,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentCommitDoc {
    pub sha: String,
    pub message: String,
    pub author_user_id: Option<String>,
    pub author_at: DateTime<Utc>,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentPageDoc {
    pub page_link: String,
    pub title: String,
    pub content: String,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentIngestedDoc {
    pub id: String,
    pub title: String,
    pub body: String,
    pub link: Option<String>,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentCode {
    pub git_url: String,
    pub commit: Option<String>,
    pub language: String,
    pub filepath: String,
    pub content: String,

    /// When start line is `None`, it represents the entire file.
    pub start_line: Option<usize>,
}

#[derive(Serialize, Deserialize)]
pub struct AttachmentClientCode {
    pub filepath: Option<String>,
    pub start_line: Option<usize>,
    pub content: String,
}

File: ee/tabby-ui/lib/utils/attachment.ts
import { AttachmentDocItem } from '../types'

export function isAttachmentWebDoc(attachment: AttachmentDocItem) {
  return (
    attachment.__typename === 'AttachmentWebDoc' ||
    attachment.__typename === 'MessageAttachmentWebDoc'
  )
}

export function isAttachmentCommitDoc(attachment: AttachmentDocItem) {
  return (
    attachment.__typename === 'AttachmentCommitDoc' ||
    attachment.__typename === 'MessageAttachmentCommitDoc'
  )
}

export function isAttachmentPullDoc(attachment: AttachmentDocItem) {
  return (
    attachment.__typename === 'AttachmentPullDoc' ||
    attachment.__typename === 'MessageAttachmentPullDoc'
  )
}

export function isAttachmentIssueDoc(attachment: AttachmentDocItem) {
  return (
    attachment.__typename === 'AttachmentIssueDoc' ||
    attachment.__typename === 'MessageAttachmentIssueDoc'
  )
}

export function isAttachmentPageDoc(attachment: AttachmentDocItem) {
  return (
    attachment.__typename === 'AttachmentPageDoc' ||
    attachment.__typename === 'MessageAttachmentPageDoc'
  )
}

export function isAttachmentIngestedDoc(attachment: AttachmentDocItem) {
  return (
    attachment.__typename === 'AttachmentIngestedDoc' ||
    attachment.__typename === 'MessageAttachmentIngestedDoc'
  )
}

export function getAttachmentDocContent(attachment: AttachmentDocItem) {
  switch (attachment.__typename) {
    case 'MessageAttachmentWebDoc':
    case 'AttachmentWebDoc':
    case 'AttachmentPageDoc':
    case 'MessageAttachmentPageDoc':
      return attachment.content
    case 'MessageAttachmentIssueDoc':
    case 'MessageAttachmentPullDoc':
    case 'AttachmentIssueDoc':
    case 'AttachmentPullDoc':
    case 'AttachmentIngestedDoc':
    case 'MessageAttachmentIngestedDoc':
      return attachment.body
    case 'MessageAttachmentCommitDoc':
    case 'AttachmentCommitDoc':
      return attachment.message
    default:
      return ''
  }
}
